third_party/txt/src/txt/paragraph_txt.cc

```diff
@@ -1186,79 +1193,92 @@ void ParagraphTxt::UpdateLineMetrics(const SkFontMetrics& metrics,
+                                   ? style.height * style.font_size
+                                   : metrics_font_height + metrics.fLeading;
+
+    // Scale the ascent and descent such that the sum of ascent and
+    // descent is `style.height * style.font_size`.
     //
-    // Disable first line ascent modifications.
-    if (line_number == 0 && paragraph_style_.text_height_behavior &
-                                TextHeightBehavior::kDisableFirstAscent) {
-      ascent = -metrics.fAscent;
-    }
-    // Disable last line descent modifications.
-    if (line_number == line_limit - 1 &&
-        paragraph_style_.text_height_behavior &
-            TextHeightBehavior::kDisableLastDescent) {
-      descent = metrics.fDescent;
-    }
+    // The raw metrics do not add up to fontSize. The state of font
+    // metrics is a mess:
+    //
+    // Each font has 4 sets of vertical metrics:
+    //
+    // * hhea: hheaAscender, hheaDescender, hheaLineGap.
+    //     Used by Apple.
+    // * OS/2 typo: typoAscender, typoDescender, typoLineGap.
+    //     Used sometimes by Windows for layout.
+    // * OS/2 win: winAscent, winDescent.
+    //     Also used by Windows, generally will be cut if extends past
+    //     these metrics.
+    // * EM Square: ascent, descent
+    //     Not actively used, but this defines the 'scale' of the
+    //     units used.
+    //
+    // `Use Typo Metrics` is a boolean that, when enabled, prefers
+    // typo metrics over win metrics. Default is off. Enabled by most
+    // modern fonts.
+    //
+    // In addition to these different sets of metrics, there are also
+    // multiple strategies for using these metrics:
+    //
+    // * Adobe: Set hhea values to typo equivalents.
+    // * Microsoft: Set hhea values to win equivalents.
+    // * Web: Use hhea values for text, regardless of `Use Typo Metrics`
+    //     The hheaLineGap is distributed half across the top and half
+    //     across the bottom of the line.
+    //   Exceptions:
+    //     Windows: All browsers respect `Use Typo Metrics`
+    //     Firefox respects `Use Typo Metrics`.
+    //
+    // This pertains to this code in that it is ambiguous which set of
+    // metrics we are actually using via SkFontMetrics. This in turn
+    // means that if we use the raw metrics, we will see differences
+    // between platforms as well as unpredictable line heights.
+    //
+    // A more thorough explanation is available at
+    // https://glyphsapp.com/tutorials/vertical-metrics
+    //
+    // Doing this ascent/descent normalization to the EM Square allows
+    // a sane, consistent, and reasonable "blob_height" to be specified,
+    // though it breaks with what is done by any of the platforms above.
+    const bool shouldNormalizeFont = style.has_height_override;
+    const double font_height =
+        shouldNormalizeFont ? style.font_size : metrics_font_height;
+
+    const size_t text_height_behavior =
```

Comment:
Defaults to `ParagraphStyle.textHeightBehavior` if `TextStyle.textHeightBehavior` is not specified. 

In the framework `TextPainter` does not know about `TextStyle.height` but currently has a `textHeightBehavior` field. With this implementation `TextPainter.textHeightBehavior` has the lowest precedence in the `textHeightBehavior` defaulting rules.

https://github.com/flutter/flutter/blob/b926c7b696eb1a5f601ef74cd191b3b6b2b22729/packages/flutter/lib/src/painting/text_painter.dart#L411-L437